### ISSUE TITLE
# Release 4.3.3?

### ISSUE THREAD

**Issue by mloy** [zeromq/libzmq#3779]
*Please use this template for reporting suspected bugs or requests for help.*

# Issue description

# Environment

* libzmq version (commit hash if unreleased): 
* OS: 

# Minimal test code / Steps to reproduce the issue

1.  

# What's the actual result? (include assertion message & call stack if applicable)

# What's the expected result?

**Comment by mloy** [comment]
I'd like to ask whether there will be a release soon? We would like to use communication via websocket, which is available on master.

Regards,
Matthias

**Comment by brettviren** [comment]
+1 

I mentioned it in IRC but to put it here, I hope CLIENT/SERVER can leave draft in the next release.  There were three open issues with the CLIENT/SERVER label.  One just got closed and there are two more still open.  Here's my amateur analysis:

#3248 appears blocked by a philosophical issue.  @somdoron, in IRC it was suggested that you might weigh in on this.  As I understand it, current sockets, including CLIENT will accept a message even if not yet connected.  This behavior is counter to statements in the RFC.  Non-draft sockets should not be changed as user code likely relies on this behavior.  The choice then, as I see it: keep CLIENT as-is (behavior consistent with other socket types) or modify CLIENT to match RFC.  If the first option is acceptable this issue could be closed.  Otherwise, CLIENT needs to stay in draft until made RFC-compliant.  FWIW, I prefer the former (ok as-is).

#2941 also needs a @somdoron mention.  As I understand the issue, a feature must be added to expose the FD in order to allow thread-safe sockets to participate in FD-based event loops.  If left open, does this issue need to block moving CLIENT/SERVER out of draft status?

**Comment by somdoron** [comment]
I will take a look tomorrow. @brettviren  are you coming to zeromq hackathon in the beginning and of February? We can work on those issues

**Comment by somdoron** [comment]
@mloy I will try to finalize the WebSocket transport during the hackathon.  It is almost ready...

**Comment by bluca** [comment]
Before the draft sockets become stable, the poller interface needs to be stable too, otherwise they aren't really usable. There are more issues open with that, including a pending pull request. Also, they need to be documented in the RFC.

**Comment by brettviren** [comment]
@somdoron I won't attend, but I'll try to help nonetheless.  So far my exposure to libzmq internals is rather limited.

**Comment by mloy** [comment]
Thank you for the very promising answers!

After writing this issue I pulled the current master to run my program with the current version. Unfortunately websocket upoort does not work anymore. 
Commit c711941e9a0cccbad8a1a74beaa0362a787d7d23 did something bad for my case. My test program uses a websocket without a path. This does not work anymore.

I made a small pull request (#3781) to fix the problem.

**Comment by somdoron** [comment]
@mloy are you using jszmq on the client side?

**Comment by mloy** [comment]
@somdoron I'm going to write a tests. I did not use jszmq, I wrote a simple pub/sub program. It uses the same logic for several types of connection (websocket, tcp, ipc and inproc).

**Comment by bluca** [comment]
I compiled the changelog https://github.com/zeromq/libzmq/pull/3784

**Comment by mjvk** [comment]
Any news on this? There are quite a few fixes (not features) that are well worth a release in my opinion.

**Comment by fndisme** [comment]
Have any news?  Bug fixed or new feathers?

**Comment by Critical2104** [comment]
Any news on this concern? It would be nice to have some fixes brought in and also CLIENT/SERVER would be a very welcome add to the library.

**Comment by honglei** [comment]
Any new? how about ZMQ_ROUTER_NOTIFY?

**Comment by somdoron** [comment]
Not sure about the version yet, but checkout ZMQ_HELLO_MSG & ZMQ_DISCONNECT_MSG

**Comment by honglei** [comment]
@somdoron   
1) ZMQ_DISCONNECT_MSG is not worked for DEALER socket， and it is not available in libzmq4.3.2. Users that use dealer-router mode wanna know disconnection on both **dealer** side and router side.

2) The last stable version released ten months ago. Pyzmq I used in pypi still builded with libzmq4.3.2. So, python users  need to build themselves if  want to use these new features, which is not easy for python users.
```python
zmq.zmq_version()
#'4.3.2'
zmq.__version__
#'19.0.1'
```

**Comment by somdoron** [comment]
I can make ZMQ_DISCONNECT_MSG on dealer as well, we are working on the next release.

**Comment by honglei** [comment]
Any news?

**Comment by TobiSchluter** [comment]
How's it going with the release? Is it time for a maintenance release model where releases will be cut whether new features are included or not? I've been holding some development, waiting for a release that includes my pull request [1] for a year now and I see that at least one zmq project is working around the issue that it fixed [2], so a new release may reduce your maintenance burden as well  ☺️

[1] https://github.com/zeromq/libzmq/pull/3681
[2] https://github.com/zeromq/cppzmq/pull/391#issuecomment-599513948

**Comment by honglei** [comment]
It's really a bad idea for not release any maintenace version for more than one year.  This issue has been opened nearly 9 months.  People should reconsider whether it is still a good idea of using libzmq in products.

**Comment by TobiSchluter** [comment]
Yay, 4.3.3 was released. The oldest bug is no longer #110 but #989. So I would say that's a lot of progress 👍
